### DISCUSSION TITLE
How can I simplify polynomials to a canonical form?

### DISCUSSION THREAD

**Discussion by t0mdd** [josdejong/mathjs#3126] [Q&A]
My application needs a way to extract the coefficients from a polynomial string after it has been simplified, math.simplify is producing some peculiar looking results:
1 - z^4 + z^5 - z^6 + z^2 - 7z^2 - z^8 simplifies to 1 + z ^ 5 + -z ^ 8 - 6 * z ^ 2 - (z ^ 4 + z ^ 6)

I am not sure what else it could look like so I am not comfortable using it just yet. I would prefer a more predictable form where the coefficients can be easily extracted, something like 1 + z ^ 5 - z ^ 8 - 6 * z ^ 2 - z ^ 4 - z ^ 6 would be better. 

For general polynomials something like each term satisfying the regex 
(coefficient)?(\\*)?(variable(\\^number)?)? and joined by + would be good. Is there some way of doing this or if not would you consider adding something like this? If so I would be willing to help.

Thanks,
Tom

**Comment by josdejong** [comment]
Mathjs doesn't have a ready made function to normalize a polynomial. It would be an interesting addition.

It could be interesting to see if we can improve `simplify` to do this better out of the box. Right now you can alter the rules that `simplify` uses yourself, so you could play around with that and come up with a set of rules that can be used to normalize a polynomial. It is also possible to parse the expression and then transform it yourself, applying rules needed to normalize the expression.

Docs:
- https://mathjs.org/docs/reference/functions/simplify.html
- https://mathjs.org/docs/expressions/expression_trees.html → search for `transform`
